File: data_preprocessing/lib_external/biosignal/camera/python/resp.py
import cv2
import numpy as np
from scipy import signal
from argparse import ArgumentParser

import cam


def preprocess(image, blur=0):
    H, W, C = image.shape
    image = cv2.resize(image, (W//2, H//2))
    image = image[H//4:]
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    if blur:
        # Aliasing 현상을 개선하기 위한 부분입니다. (experimental)
        gray = cv2.medianBlur(gray, blur)
        # gray = cv2.blur(gray, (blur, blur))
    return image, gray


class FlowModule:
    """
    Optical Flow를 계산하기 위한 클래스입니다.
    두 이미지가 필요하므로 이전 프레임 이미지를 prev에 기억합니다.
    https://docs.opencv.org/3.4/d4/dee/tutorial_optical_flow.html
    """

    def __init__(self):
        self.prev = None
        self.next = None

    def __call__(self, image):
        self.prev = self.next
        self.next = image

        if self.prev is None:
            # frame이 1번만 입력되었을 때 zeros array를 리턴합니다.
            flow = np.zeros((image.shape[0], image.shape[1], 2), np.float32)
        else:
            flow = cv2.calcOpticalFlowFarneback(
                self.prev, self.next, None, 0.5, 3, 15, 3, 5, 1.2, 0)

        return flow

    @staticmethod
    def get_flow_view(flow):
        # Optical Flow 시각화 함수입니다.
        h, w, c = flow.shape
        hsv = np.zeros((h, w, 3), dtype=np.uint8)
        mag, ang = cv2.cartToPolar(flow[..., 0], flow[..., 1])
        hsv[..., 0] = ang*90/np.pi
        hsv[..., 1] = 255
        hsv[..., 2] = cv2.normalize(mag, None, 0, 255, cv2.NORM_MINMAX)
        bgr = cv2.cvtColor(hsv, cv2.COLOR_HSV2BGR)
        return bgr


class Queue:
    # 리스트의 사이즈를 고정하기 위한 클래스입니다.
    def __init__(self, size):
        self.size = size
        self.data = [0.0 for i in range(self.size)]

    def __len__(self):
        return len(self.data)

    def __call__(self, x):
        self.data.append(x)
        if len(self.data) > self.size:
            self.data.pop(0)

    def __getitem__(self, idx):
        return self.data[idx]


class SignalModule:
    # 호흡 신호를 처리하기 위한 클래스입니다.
    def __init__(self, size=360):
        self.size = size
        self.reset()

    def reset(self):
        self.times = Queue(self.size)
        self.pq = Queue(self.size)
        self.iq = Queue(self.size)
        self.integral = 0.0
        self.frequency = 0.0

    def __call__(self, timestamp, value):
        value = np.clip(value, -0.5, 0.5)
        self.times(timestamp)
        self.pq(value)
        self.integral += value
        self.iq(self.integral)

        # process
        sig = np.asarray(self.iq.data)
        # sig = signal.detrend(sig)

        # find_peaks로 들숨과 날숨 시점을 찾습니다.
        # TODO: parameter tuning
        inhalation, _ = signal.find_peaks(sig, prominence=0.05, width=5)
        exhalation, _ = signal.find_peaks(-sig, prominence=0.05, width=5)

        self.inhalation = inhalation
        self.exhalation = exhalation

        # Respiration Rate를 계산합니다. (Hz)
        if len(exhalation) >= 2:
            p0 = exhalation[0]
            p1 = exhalation[-1]
            t0 = self.times[p0]
            t1 = self.times[p1]
            dt = t1 - t0
            count = len(exhalation)
            frequency = (count - 1) / dt
            self.frequency = frequency

    def get_graph(self):
        # 신호 그래프 시각화 부분입니다.
        scale = 320
        v = (np.asarray(self.pq.data) + 0.5) * scale
        v = np.clip(v, 0, scale-1).astype(int)
        size = len(v)
        sig = np.asarray(self.iq.data)
        s = (sig - sig.min()) / (sig.max() - sig.min() + 1e-6)
        s = (s * scale).astype(int)

        graph = np.zeros((size, scale, 3), dtype=np.uint8)
        for peak in self.inhalation:
            graph[peak, :, 0] = 255
        for peak in self.exhalation:
            graph[peak, :, 2] = 255

        def reorder(p0, p1):
            if p0 == p1:
                return p0, p1+1
            elif p0 > p1:
                return p1, p0
            return p0, p1

        for i in range(1, size):
            p0, p1 = reorder(v[i-1], v[i])
            graph[i, p0:p1, 0:2] = 128
            p0, p1 = reorder(s[i-1], s[i])
            graph[i, p0:p1, 1] = 255
        return graph


def main(
    recording=False,
    blur=0,
):
    cv2.namedWindow('view')
    cv2.moveWindow('view', 0, 0)
    cv2.namedWindow('input')
    cv2.moveWindow('input', 0, 400)

    camera = cam.Cam(0)
    f_module = FlowModule()
    s_module = SignalModule()

    if recording:
        rec = cam.Rec('./records', 15.0, (960, 360))

    while camera():
        image = camera.frame
        timestamp = camera.time

        # 이미지 전처리 부분입니다.
        # Resize -> Crop -> Gray
        frame, gray = preprocess(image, blur)
        cv2.imshow('input', gray)

        # optical flow를 계산합니다.
        flow = f_module(gray)

        # y 방향의 flow의 평균값을 타임스탬프와 함께 signal module에 입력합니다.
        value = flow[..., 1].mean()
        s_module(timestamp, value)

        # Visualization
        flow_view = f_module.get_flow_view(flow)
        graph = s_module.get_graph()
        view = np.concatenate([frame, flow_view], axis=0)
        view = np.concatenate([view, graph], axis=1)

        logs = [
            ' %4.1f FPS' % camera.fps,
            ' %4.1f brpm' % (s_module.frequency * 60)
        ]
        if recording:
            logs.append(' Rec.')
        color = (0, 255, 255)
        org = [640, 40]
        for log in logs:
            cv2.putText(view, log, org, 0, 0.5, color, 1)
            org[1] += 20

        cv2.imshow('view', view)
        if recording:
            rec(view)


if __name__ == '__main__':
    parser = ArgumentParser()
    parser.add_argument('--recording', type=int, default=0)
    args = parser.parse_args()
    main(recording=args.recording)

File: data_preprocessing/lib_external/biosignal/camera/cpp/app/resp.cc
#include <algorithm>
#include <cstdio>
#include <iostream>
#include <vector>

#include <opencv2/opencv.hpp>
#include <opencv2/videoio.hpp>

#include "cam.h"
#include "signalgo.h"

class FlowModule {
  cv::Mat prev;
  cv::Mat next;
  int count;

public:
  FlowModule();

  cv::Mat run(cv::Mat image);
};

FlowModule::FlowModule() { count = 0; }

cv::Mat FlowModule::run(cv::Mat image) {
  prev = next;
  next = image;

  cv::Mat flow;
  count++;
  if (count >= 2) {
    flow = cv::Mat(image.size(), CV_32FC2);
    cv::calcOpticalFlowFarneback(prev, next, flow, 0.5, 3, 15, 3, 5, 1.2, 0);
  } else {
    flow = cv::Mat::zeros(image.size(), CV_32FC2);
  }
  return flow;
}

cv::Mat GetFlowView(cv::Mat flow) {
  cv::Mat flow_parts[2];
  cv::split(flow, flow_parts);
  cv::Mat magnitude, angle, magn_norm;
  cv::cartToPolar(flow_parts[0], flow_parts[1], magnitude, angle, true);
  cv::normalize(magnitude, magn_norm, 0.0f, 1.0f, cv::NORM_MINMAX);
  angle *= ((1.f / 360.f) * (180.f / 255.f));

  cv::Mat _hsv[3], hsv, hsv8, bgr;
  _hsv[0] = angle;
  _hsv[1] = cv::Mat::ones(angle.size(), CV_32F);
  _hsv[2] = magn_norm;
  cv::merge(_hsv, 3, hsv);
  hsv.convertTo(hsv8, CV_8U, 255.0);
  cv::cvtColor(hsv8, bgr, cv::COLOR_HSV2BGR);

  return bgr;
}

// TODO: circular queue?
class Queue {
  unsigned int size_;

public:
  std::vector<double> data;
  Queue();
  Queue(int size);
  void set(int size);
  void push_back(double x);
  int size() { return size_; };
};

Queue::Queue() { size_ = 0; }

void Queue::set(int size) {
  size_ = size;
  for (int i = 0; i < size; i++) {
    data.push_back(0.f);
  }
}

void Queue::push_back(double x) {
  data.push_back(x);
  if (data.size() > size_) {
    data.erase(data.begin());
  }
}

double clip(double value, double min, double max) {
  if (value < min) {
    return min;
  } else if (value > max) {
    return max;
  }
  return value;
}

class SignalModule {
  int size_;
  double integral;

public:
  Queue times;
  Queue pq;
  Queue iq;
  std::vector<int> inspiration;
  double frequency;
  SignalModule(int size);
  void run(double timestamp, double value);
};

SignalModule::SignalModule(int size) {
  size_ = size;
  times.set(size);
  pq.set(size);
  iq.set(size);
  integral = 0.f;
  frequency = 0.f;
}

void SignalModule::run(double timestamp, double value) {
  value = clip(value, -0.5, +0.5);

  times.push_back(timestamp);
  pq.push_back(value);
  integral += value;
  iq.push_back(integral);

  inspiration = signalgo::find_peaks(iq.data, 0.05, -1, 5);

  int count = inspiration.size();
  if (count >= 2) {
    int p0 = inspiration[0];
    int p1 = inspiration[inspiration.size() - 1];
    double t0 = times.data[p0];
    double t1 = times.data[p1];
    double dt = t1 - t0;
    frequency = (count - 1) / dt;
  }
}

cv::Mat get_graph(SignalModule sm) {
  const int scale = 320;
  const int size = sm.pq.size();
  cv::Mat graph = cv::Mat::zeros(cv::Size(scale, size), CV_8UC3);

  for (int i = 0; i < (int)sm.inspiration.size(); i++) {
    int r = sm.inspiration[i];
    for (int x = 0; x < scale; x++) {
      graph.data[(r * scale + x) * 3] = 255; // red channel;
    }
  }

  const double min =
      *std::min_element(std::begin(sm.iq.data), std::end(sm.iq.data));
  const double max =
      *std::max_element(std::begin(sm.iq.data), std::end(sm.iq.data));
  const double max_min = max - min + 1e-6;

  int s_prev = (int)((sm.iq.data[0] - min) / max_min * scale);
  for (int i = 0; i < size; i++) {
    double v = sm.iq.data[i];
    v = (v - min) / max_min;
    int s = (int)(v * scale);

    int p0 = std::min(s, s_prev);
    int p1 = std::max(s, s_prev);
    for (int p = p0; p < p1; p++) {
      graph.data[(i * scale + p) * 3 + 1] = 255; // green channel;
    }
    s_prev = s;
  }

  return graph;
}

int test() {
  cv::namedWindow("view");
  cv::moveWindow("view", 0, 0);
  cv::namedWindow("input");
  cv::moveWindow("input", 0, 400);

  cam::Cam camera(0);
  int width = camera.get_width() / 2;
  int height = camera.get_height() / 2;

  auto f_module = FlowModule();
  auto s_module = SignalModule(360);

  cv::Mat image, image_p, gray;

  while (camera.read()) {
    auto frame = camera.get_frame();
    cv::resize(frame, image, cv::Size(width, height));

    image_p = image(cv::Rect(0, height / 2, width, height / 2));
    cv::cvtColor(image_p, gray, cv::COLOR_BGR2GRAY);
    cv::imshow("input", gray);

    cv::Mat flow = f_module.run(gray.clone());
    cv::Mat flow_view = GetFlowView(flow);

    auto mean = cv::mean(flow);

    s_module.run(camera.get_time(), mean[1]);

    cv::Mat graph = get_graph(s_module);

    cv::Mat view, view2;
    cv::vconcat(image_p, flow_view, view);
    cv::hconcat(view, graph, view2);

    cv::Point pos(640, 40);
    cv::Scalar yellow(0, 255, 255);
    char buf[64] = {0};
    sprintf(buf, " %4.1f fps", camera.get_fps());
    cv::putText(view2, buf, pos, 0, 0.5, yellow, 1);
    pos.y += 20;
    sprintf(buf, " %4.1f brpm", s_module.frequency * 60);
    cv::putText(view2, buf, pos, 0, 0.5, yellow, 1);

    cv::imshow("view", view2);
  }

  return 0;
}

int main() {
  test();
  return 0;
}